### ISSUE TITLE
Memory leak in `_contexts`

### ISSUE THREAD

**Issue by holm** [Jeff-Lewis/cls-hooked#11]
We upgraded to Node 8 yesterday, and are now seeing memory leaks. Taking some heap snapshots I can see the `_contexts` map growing all the time. We run 4.2.2, so it shouldn't be the memory leak fixed recently.

It's hard to follow the code, but I would guess this is related to how the promises are handled. At least it seems the contexts are not always removed from the map.

**Comment by holm** [comment]
I looked into this further, and it seems destroy is just never called for some asyncIds. Reading other issues from Node it seems this is caused by the promises not being destroyed until GC. However for some reason they seem to never be gc'ed, and I am not sure how to debug that.

**Comment by Jeff-Lewis** [comment]
I'm looking into something we can do with the context's `enter()` and `exit()` and remove items from _context `exit()`. Not sure yet if that's the answer.

**Comment by holm** [comment]
There is a similar looking issue in Node, see https://github.com/nodejs/node/issues/14446#issuecomment-317673776. This looks from a quick glance to be the same issue.

**Comment by clakech** [comment]
Hi, any news about this issue, what should be done in the code of cls-hooked to solve this one ?

**Comment by clakech** [comment]
How seems like, it is a normal behavior in this issue https://github.com/nodejs/node/issues/14446#issuecomment-317673776 

so maybe this issue here is OK too.

**Comment by holm** [comment]
We eventually worked out what was causing this. Turned out to actually be the NewRelic that was the main culprit, albeit we also had some of our own.

The fact that the contexts are only destroyed when the promise is garbage collected, requires you to be very careful that you do not in some indirect way reference the promises from the data you store in the cls.
